libs/ballot-interpreter-nh/src/scoring.rs

```diff
@@ -238,3 +239,103 @@ pub fn score_bubble_mark(
         fill_diff_image: diff_image,
     })
 }
+
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ScoredPositionArea {
+    pub grid_position: GridPosition,
+    pub bounds: Rect,
+    pub score: UnitIntervalScore,
+}
+
+pub type ScoredPositionAreas = Vec<ScoredPositionArea>;
+
+/// Computes scores for all the write-in areas in a scanned ballot image. This could
+/// be used to determine which write-in areas are most likely to contain a write-in
+/// vote even if the bubble is not filled in.
+pub fn score_write_in_areas(
+    img: &GrayImage,
+    grid_layout: &GridLayout,
+    contest_layouts: &[InterpretedContestLayout],
+    scored_bubble_marks: &ScoredBubbleMarks,
+    debug: &ImageDebugWriter,
+) -> Vec<ScoredPositionArea> {
+    let threshold = otsu_level(img);
+
+    let scored_write_in_areas = grid_layout
+        .write_in_positions()
+        .iter()
+        .filter_map(|grid_position| {
+            let (_, scored_bubble_mark) =
+                scored_bubble_marks
+                    .iter()
+                    .find(|(location, scored_bubble_mark)| {
+                        scored_bubble_mark.is_some()
+                            && location.location() == grid_position.location()
+                    })?;
+            let scored_bubble_mark = scored_bubble_mark.as_ref()?;
+            score_write_in_area(
+                img,
+                grid_position,
+                scored_bubble_mark.matched_bounds,
+                contest_layouts,
+                threshold,
+            )
+        })
+        .collect();
+
+    debug.write("scored_write_in_areas", |canvas| {
+        debug::draw_scored_write_in_areas(canvas, &scored_write_in_areas);
+    });
+
+    scored_write_in_areas
+}
+
+fn score_write_in_area(
+    img: &GrayImage,
+    grid_position: &GridPosition,
+    scored_bubble_mark_bounds: Rect,
+    contest_layouts: &[InterpretedContestLayout],
+    threshold: u8,
+) -> Option<ScoredPositionArea> {
+    let GridPosition::WriteIn { side, column, row, .. } = *grid_position else {
+        return None;
+    };
+    let contest_layout = contest_layouts
+        .iter()
+        .find(|contest_layout| contest_layout.contest_id == grid_position.contest_id())?;
+    let write_in_layout = contest_layout.options.iter().find(|option| {
+        option.grid_location.side == side
+            && option.grid_location.column == column
+            && option.grid_location.row == row
+    })?;
+    let write_in_layout_width = write_in_layout.bounds.width();
+
+    // 75% of the width of the write-in crop area and 2x the height of the
+    // bubble turned out to work pretty well with the NH state ballot layout.
+    // For now we're not trying to be more general than that.
```

Comment:
Seems fine for now. Do you have any ideas how we might solve this in the more general case?

Reply:
I think we'll need to specify per-contest or maybe even per-option what the corresponding offset/area from the bubble is that we should inspect. See [this Slack thread for the most recent context](https://votingworks.slack.com/archives/CEL6D3GAD/p1689021283637969).